src/systems/physics.rs

```diff
@@ -1,30 +1,46 @@
 use amethyst::derive::SystemDesc;
-use amethyst::ecs::{Read, System, SystemData, ReadStorage, WriteStorage};
+use amethyst::ecs::{Read, System, SystemData, ReadStorage, WriteStorage, Entities};
 use amethyst::ecs::prelude::Join;
 use amethyst::core::{Transform, Time};
+use amethyst::core::math::Vector2;
 
-use crate::components::Velocity;
+use crate::components::{Velocity, DistanceLimit};
 
 #[derive(SystemDesc)]
 pub struct PhysicsSystem;
 
 impl<'s> System<'s> for PhysicsSystem {
     type SystemData = (
+        Entities<'s>,
         WriteStorage<'s, Transform>,
         ReadStorage<'s, Velocity>,
-        Read<'s, Time>
+        Read<'s, Time>,
+        WriteStorage<'s, DistanceLimit>
     );
 
-    fn run(&mut self, (mut transforms, velocities, time): Self::SystemData) {
-        for (transform, velocity) in (&mut transforms, &velocities).join() {
-            update_position(transform, velocity, &time);
+    fn run(&mut self, (entities, mut transforms, velocities, time, mut distance_limits): Self::SystemData) {
+        for (moving_entity, transform, velocity, maybe_distance_limit) in
+            (&*entities, &mut transforms, &velocities, (&mut distance_limits).maybe()).join()
+        {
+            let delta_s = update_position(transform, velocity, &time).norm();
+
+            if let Some(distance_limit) = maybe_distance_limit {
```

Comment:
Strictly speaking distance limitation should not be in the same system as position update. (One of them accesses transform and velocity but not distance limit). And there is no data dependency between position update and distance limit handling. For now it's okay, but I think it should be separated later.

Reply:
There is a dependency right now, the "delta_s" / traveled distance. But it can be extracted by  storing previous position in DistanceLimit, and do the diffing in a separate system.

Reply:
Oh, I see. I missed it. This is OK then.
(Because of some reason. I thought that distance is calculated as time)

Reply:
Actually, in case of a bullet, which travels in a straight line with a known fixed speed, traveled distance can be calculated based on time, so for example a TimeLimit component instead of DistanceLimit. TimeLimit could have other uses as well (health packs, dropped weapons disappear after a few minutes, etc.)